### PR TITLE
VersionedLayerClient::StartBatch use sync repo

### PR DESCRIPTION
**Pull request by kzvieriev**
Refactor StartBatch code to make it clear. Also added unit and
integration tests.

Resolves: OLPEDGE-1409

Signed-off-by: Kostiantyn Zvieriev <ext-kostiantyn.zvieriev@here.com>

### CODE REVIEW

File: olp-cpp-sdk-dataservice-write/tests/CMakeLists.txt [hunk 1]
```diff
@@ -23,6 +23,7 @@ set(OLP_SDK_DATASERVICE_WRITE_TEST_SOURCES
     StartBatchRequestTest.cpp
     StreamLayerClientImplTest.cpp
     TimeUtilsTest.cpp
+	VersionedLayerClientImplTest.cpp
```
Comment (andescu) [inline olp-cpp-sdk-dataservice-write/tests/CMakeLists.txt:26]: Wrong indentation.

File: tests/integration/olp-cpp-sdk-dataservice-write/VersionedLayerClientTest.cpp [hunk 2]
```diff
@@ -0,0 +1,310 @@
...
+    SCOPED_TRACE("Success request");
+
+    EXPECT_CALL(*network_, Send(IsGetRequest(kLookupPublishApiUrl), _, _, _, _))
+        .WillOnce(ReturnHttpResponse(olp::http::NetworkResponse().WithStatus(
+                                         olp::http::HttpStatusCode::OK),
+                                     olp::serializer::serialize(api)));
+    EXPECT_CALL(*network_, Send(IsPostRequest(kPublishUrl), _, _, _, _))
+        .WillOnce(ReturnHttpResponse(olp::http::NetworkResponse().WithStatus(
+                                         olp::http::HttpStatusCode::OK),
+                                     olp::serializer::serialize(publication)));
```
Comment (andescu) [inline tests/integration/olp-cpp-sdk-dataservice-write/VersionedLayerClientTest.cpp:111]: I wonder. Is there no token request before?

File: tests/integration/olp-cpp-sdk-dataservice-write/VersionedLayerClientTest.cpp [hunk 3]
```diff
@@ -0,0 +1,310 @@
...
+
+    write::VersionedLayerClient write_client(kHrn, settings_);
+    const auto batch_request = model::StartBatchRequest().WithLayers({kLayer});
+    const auto response =
+        write_client.StartBatch(batch_request).GetFuture().get();
+    const auto& result = response.GetResult();
```
Comment (andescu) [inline tests/integration/olp-cpp-sdk-dataservice-write/VersionedLayerClientTest.cpp:117]: Please separate a little request and waiting for the response.  Mixing it like this together makes it very hard to read.

File: tests/integration/olp-cpp-sdk-dataservice-write/VersionedLayerClientTest.cpp [hunk 4]
```diff
@@ -0,0 +1,310 @@
...
+
+TEST_F(VersionedLayerClientTest, StartBatch) {
+  const auto catalog = kHrn.ToCatalogHRNString();
+  const auto api = CreateApiResponse(kPublishApiName);
+  const auto publication =
+      mockserver::DefaultResponses::GeneratePublicationResponse({kLayer}, {});
+  ASSERT_FALSE(api.empty());
+
+  {
+    SCOPED_TRACE("Success request");
```
Comment (andescu) [inline tests/integration/olp-cpp-sdk-dataservice-write/VersionedLayerClientTest.cpp:102]:
```suggestion
    SCOPED_TRACE("Successful request, future");
```

File: tests/integration/olp-cpp-sdk-dataservice-write/VersionedLayerClientTest.cpp [hunk 5]
```diff
@@ -0,0 +1,310 @@
...
+    ASSERT_TRUE(result.GetDetails());
+    ASSERT_TRUE(result.GetLayerIds());
+    ASSERT_EQ(result.GetLayerIds().get().size(), 1);
+    ASSERT_EQ(result.GetLayerIds().get().front(), kLayer);
+    ASSERT_NE("", response.GetResult().GetId().value());
+    Mock::VerifyAndClearExpectations(network_.get());
+  }
+
+  {
+    SCOPED_TRACE("Success request, but with callback");
```
Comment (andescu) [inline tests/integration/olp-cpp-sdk-dataservice-write/VersionedLayerClientTest.cpp:130]:
```suggestion
    SCOPED_TRACE("Successful request, callback");
```

File: tests/integration/olp-cpp-sdk-dataservice-write/VersionedLayerClientTest.cpp [hunk 6]
```diff
@@ -0,0 +1,310 @@
...
+
+    // don't mock lookup api request since its cached in previous step
+    EXPECT_CALL(*network_, Send(IsPostRequest(kPublishUrl), _, _, _, _))
```
Comment (andescu) [inline tests/integration/olp-cpp-sdk-dataservice-write/VersionedLayerClientTest.cpp:133]: Ideally, you would test here that it takes the lookup API from the cache and not requesting it again.

File: tests/integration/olp-cpp-sdk-dataservice-write/VersionedLayerClientTest.cpp [hunk 7]
```diff
@@ -0,0 +1,310 @@
...
+    auto future = promise.get_future();
+
+    write::VersionedLayerClient write_client(kHrn, settings_);
+    const auto batch_request = model::StartBatchRequest().WithLayers({kLayer});
+    const auto token = write_client.StartBatch(
+        batch_request, [&promise](write::StartBatchResponse response) {
+          promise.set_value(std::move(response));
+        });
+    const auto response = future.get();
+    const auto& result = response.GetResult();
```
Comment (andescu) [inline tests/integration/olp-cpp-sdk-dataservice-write/VersionedLayerClientTest.cpp:148]: Same here please separate the request and the waiting:

```suggestion
    std::promise<write::StartBatchResponse> promise;
    write::VersionedLayerClient write_client(kHrn, settings_);
    const auto batch_request = model::StartBatchRequest().WithLayers({kLayer});
    const auto token = write_client.StartBatch(
        batch_request, [&promise](write::StartBatchResponse response) {
          promise.set_value(std::move(response));
        });
        
     auto future = promise.get_future();   
    const auto response = future.get();
    const auto& result = response.GetResult();
```

File: tests/integration/olp-cpp-sdk-dataservice-write/VersionedLayerClientTest.cpp (hunk 2 above)
Comment (kzvieriev) [reply tests/integration/olp-cpp-sdk-dataservice-write/VersionedLayerClientTest.cpp:111]: Auth settings are `boost::none`, so no token requests in the tests.

File: tests/integration/olp-cpp-sdk-dataservice-write/VersionedLayerClientTest.cpp (hunk 6 above)
Comment (kzvieriev) [reply tests/integration/olp-cpp-sdk-dataservice-write/VersionedLayerClientTest.cpp:136]: Test will fail if lookup API request occur.
Anyway, added mock with `Times(0)` so its clear API requests are not welcome here :-)

File: tests/integration/olp-cpp-sdk-dataservice-write/VersionedLayerClientTest.cpp (hunk 2 above)
Comment (andescu) [reply tests/integration/olp-cpp-sdk-dataservice-write/VersionedLayerClientTest.cpp:111]: But why? You should test properly all the way.

Comment (kzvieriev) [reply tests/integration/olp-cpp-sdk-dataservice-write/VersionedLayerClientTest.cpp:136]: done

**Pull request merged**